src/MediatR/IMediator.cs

```diff
@@ -23,7 +24,6 @@ public interface IMediator
         /// <param name="notification">Notification object</param>
         /// <param name="cancellationToken">Optional cancellation token</param>
         /// <returns>A task that represents the publish operation.</returns>
-        Task Publish<TNotification>(TNotification notification, CancellationToken cancellationToken = default)
-            where TNotification : INotification;
+        Task Publish(INotification notification, CancellationToken cancellationToken = default);
```

Comment:
This might still require a cast on the caller side - we can mitigate this by checking to see if the object is an `INotification`, and throw an `ArgumentException` otherwise.

The overload rules are weird too - the generic method might never be called because the compiler assumes this overload.

Reply:
Please check: https://github.com/jbogard/MediatR/pull/300#discussion_r206524326

About overload - it is redundant with generic, so generic one has been removed. It should be brought back if we decide to go for `object` instead of `INotification`